### PR TITLE
Support primary keys which are not called "id"

### PR DESCRIPTION
**Pull request by EyMaddis**
Moin ;)
This PR lifts the restriction that only primary keys with the name `id` are supported in Backpex by getting the primary key from Ecto.

It can now support schemas like this:

```elixir
  @primary_key false
  schema "short_links" do
    field :short_key, :string, primary_key: true
    field :url, :string

    belongs_to :product, Demo.Product, type: :binary_id

    timestamps()
  end
```

I've added an example called "Short Links" (which are linked to products) to the demo to show it in action.

I also sneaked in a helpful error message for a problem that took me a while to debug (using `something_id` instead of `something` in a field).

This was implemented while playing around with Backpex in order to *maybe* use it with [Partey.io](https://partey.io/?mtm_campaign=backpex). 

It is

1. my first "bigger" implementation in Elixir
2. my first-ever Elixir PR
3. My first time using Backpex at all

so feel free to give feedback if anything is not feeling correct for an Elixir codebase.

This basically replaces all occurrences of `.id` with something else, except in cases where a LiveView id is referenced.

Each LiveResource now has a function called `get_primary_key_field` which can be used to directly access the primary field info.

### CODE REVIEW

**Comment by EyMaddis** [comment]
I've fixed the linting issue :)

**Comment by Flo0807** [comment]
Moin @EyMaddis! Thanks for the pull request! This will make it much easier to use Backpex in projects with different requirements 🚀 I will give you feedback on the PR by the end of the week 🙂

File: lib/backpex/live_resource.ex [hunk 1]
```diff
@@ -922,6 +925,10 @@ defmodule Backpex.LiveResource do
 
       def get_empty_filter_key, do: @empty_filter_key
 
+      def get_primary_key_field, do: @primary_key_field
+
+      defp primary_key(item), do: Map.get(item, @primary_key_field)
```
Comment (Flo0807) [inline lib/backpex/live_resource.ex:930]: We should define `primary_key(item)` and `primary_key(assigns, item)` next to each other:

```elixir
defp primary_key(item), do: Map.get(item, @primary_key_field)
defp primary_key(assigns, item), do: Map.get(item, assigns.primary_key_field)
```

Is the second function (with the assigns as the first param) even called? I cannot find any references to this function.

**Review by Flo0807 — CHANGES_REQUESTED** [review]
I noticed that it is impossible to add new Shortlinks to new or existing products. If you add a short link on the product edit or new page, you cannot save the item. I debugged it and it seems the new shortlink does not get linked with the corresponding product.

See the ouput of the changeset. Note `product_id: {"can't be blank", [validation: :required]}`

```
app-1       | changeset_debug: #Ecto.Changeset<
app-1       |   action: nil,
app-1       |   changes: %{
app-1       |     short_links: [
app-1       |       #Ecto.Changeset<
app-1       |         action: :insert,
app-1       |         changes: %{url: "1", short_key: "IZMSUxt-rW4"},
app-1       |         errors: [product_id: {"can't be blank", [validation: :required]}],
app-1       |         data: #Demo.ShortLink<>,
app-1       |         valid?: false
app-1       |       >
app-1       |     ]
app-1       |   },
app-1       |   errors: [],
app-1       |   data: #Demo.Product<>,
app-1       |   valid?: false
app-1       | >
```

File: demo/lib/demo/short_link.ex [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+  @required_fields ~w[short_key url product_id]a
+
+  def changeset(short_link, attrs, metadata \\ []) do
+    short_link
+    |> cast(attrs, @required_fields)
+    |> cast_assoc(:product, with: &Product.changeset/2)
+    |> add_short_key()
+    |> validate_required(@required_fields)
+  end
```
[the remark this answers is not in the document]
Comment (EyMaddis) [reply demo/lib/demo/short_link.ex]: I've adapted the documentation for now. The solution should be to use `URI.encode_query(value)` whenever building a routing path.
I currently cannot test this myself (see other comment)

File: demo/lib/demo_web/live/short_link_live.ex [hunk 3]
```diff
@@ -0,0 +1,50 @@
+defmodule DemoWeb.ShortLinkLive do
```
[the remark this answers is not in the document]
Comment (EyMaddis) [reply demo/lib/demo_web/live/short_link_live.ex:1]: I've added a return_to to go back to the list

File: demo/lib/demo/short_link.ex (hunk 2 above)
[the remark this answers is not in the document]
Comment (EyMaddis) [reply demo/lib/demo/short_link.ex]: Now supported and properly encoded
image

**Comment by EyMaddis** [comment]
New version :)

File: demo/priv/repo/migrations/20240824063251_create_short_links.exs [hunk 4]
```diff
@@ -0,0 +1,14 @@
+defmodule Demo.Repo.Migrations.CreateShortLinks do
+  use Ecto.Migration
+
+  def change do
+    create table(:short_links, primary_key: false) do
+      add :short_key, :string, primary_key: true
+      add :url, :string
+
+      add(:product_id, references(:products, type: :binary_id))
```
[the remark this answers is not in the document]
Comment (EyMaddis) [reply demo/priv/repo/migrations/20240824063251_create_short_links.exs]: adapted :)

File: lib/backpex/ecto/ecto_utils.ex [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+        No primary key found. Please define a primary key.
+        """
+
+      _ ->
+        raise """
+        Compound primary keys are not supported. Please use a single primary key .
+        """
+    end
+  end
+end
```
[the remark this answers is not in the document]
Comment (EyMaddis) [reply lib/backpex/ecto/ecto_utils.ex]: yeeted that into the code

File: lib/backpex/fields/has_many_through.ex [hunk 6]
```diff
@@ -43,6 +43,7 @@ defmodule Backpex.Fields.HasManyThrough do
   The field requires a [`Ecto.Schema.has_many/3`](https://hexdocs.pm/ecto/Ecto.Schema.html#has_many/3) relation with a mandatory `through` option in the main schema. Any extra column in the pivot table besides the relational id's must be mapped in the `pivot_fields` option or given a default value.
   """
 
+  alias Backpex.Ecto.EctoUtils
```
[the remark this answers is not in the document]
Comment (EyMaddis) [reply lib/backpex/fields/has_many_through.ex]: sure

File: lib/backpex/html/resource.ex [hunk 7]
```diff
@@ -114,17 +114,20 @@ defmodule Backpex.HTML.Resource do
...
+    primary_key = Map.get(item, live_resource.get_primary_key_field())
+
     assigns =
       assigns
       |> assign(:field, field)
       |> assign(:field_options, field_options)
       |> assign(:value, Map.get(item, name))
       |> assign(:type, :index)
       |> assign(:readonly, readonly)
+      |> assign(:primary_key, primary_key)
```
[the remark this answers is not in the document]
Comment (EyMaddis) [reply lib/backpex/html/resource.ex]: fixed

File: lib/backpex/live_resource.ex (hunk 1 above)
Comment (EyMaddis) [reply lib/backpex/live_resource.ex:939]: aaaand done

**Comment by EyMaddis** [comment]
linting issues in demo fixed as well

**Review by Flo0807 — APPROVED** [review]

**Comment by Flo0807** [comment]
Thanks again for the contribution! 🎉 @EyMaddis 

I made some minor changes:

- Remove `prerequisites.md`: We recently moved these instructions to `installation.md`
- Improve validation of short_links
  - Ensure `short_link` is URL-safe
  - Ensure `short_link` is unique (also added a unique index to the migration)
- Add prompt to product select field when creating or updating short links since the field is nullable
- Update `Backpex.Fields.HasMany`: I recently refactored the field and needed to adjust it to fetch the primary key from the live resource

**Comment by EyMaddis** [comment]
You're welcome :) 

(It might actually be a good idea to have something that has invalid URL segments to test that the automatic URL encoded works :) )

**Pull request merged**